### ISSUE TITLE
Bug: [prefer-optional-chain] invalid fix with array access by index

### ISSUE THREAD

**Issue by Hypnosphi** [typescript-eslint/typescript-eslint#11917]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyARnFAFzIgCuAtsdANwEC%2BBBM1ICYw6IyWnFAAKGJkoYsAHyrUANvIA0ycegDOkzAG0AugEpKxTPIhwBeQsmAxkYzMgBkj1ZgB0pKMgC8v1xu0ABl0Afg8yfRwCIiIoCDBqKAEwKGoIaOQ2KziEpNU4eXV0-BYgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

### Repro Code

```TypeScript
interface Foo {
  bar: number;
}

function main(foo: Foo | null, foos: Foo[]): boolean {
  if (foo && foo.bar === foos[0]?.bar) {
    return true
  }
  return false
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-optional-chain": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

No error or autofix

### Actual Result

Autofix provides an invalid result:

```ts
function main(foo: Foo | null, foos: Foo[]): boolean {
  if (foo?.bar === foos[0]?.bar) {
    return true
  }
  return false
}
```

`main(null, [])` was `false` before the fix but now it's `true`

### Additional Info

Happens after upgrading to `8.51.0`, probably introduced with https://github.com/typescript-eslint/typescript-eslint/pull/11835

**Comment by bradzacher** [comment]
You've stated that this is due to array index but it's completely unrelated to that - it has nothing to do with the right-hand side of the comparison.

Instead your code has assigned meaning to the differentiation between `null` and `undefined` in the left-hand side of the comparison. So converting your left-hand side expression to an optional chain removes the differentiation between the two values, leading to the change in behaviour.

The rule, however, has no way of exactly knowing if you meant to assign different meaning to `null` and `undefined` or not - which is why it reports on the expression - to tell you that this _could_ be made an optional chain.

In general you should avoid applying different meaning to the two values because between falsey coercion, nullish coalescing, and optional chaining the language wants you to treat them the same...

---

All that being said - the rule should probably report a suggestion fixer here instead of an autofixer. I.e. If the report is in a binary expression then it should use a suggestion fixer.

**Comment by Hypnosphi** [comment]
> Instead your code has assigned meaning to the differentiation between null and undefined

@bradzacher can you please elaborate on that? In my understanding, my original code produces `false` for both `main(null, [])` and `main(undefined, [])` because `foo` is falsy. And the autofixed version produces `true` in both cases.

The reason I think it has something to do with array index access is that there's no error/fix with this code:
https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyARnFAFzIgCuAtsdANwEC%2BBBM1ICYw6IyWnFAAKGJkoYsAHyrUANvIA0JMpMzJZNRQEpKxTPIhwBeQsmAxkYjQDJbycegB0pKMgC8X1VAD8rsh0cAiIiKAgwaigBMChqCBDkNnNwyOjHOHkAZwT8FiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false
```
interface Foo {
  bar: number;
}

function main(foo: Foo | null, bar: Foo | null): boolean {
  if (foo && foo.bar === bar?.bar) {
    return true
  }
  return false
}
```

In fact, initially my example didn't even have optional chaining, I just applied the first autofix here which is legitimate:

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyARnFAFzIgCuAtsdANwEC%2BBBM1ICYw6IyWnFAAKGJkoYsAHyrUANvIA0ycegDOkzAG0AugEpKxTPIhwBeQsmAxkYzMgBkj1ZnXaADLqcu1AOlIoZABeUNcNT10Asn0cAiIiKAgwaigBMChqCHjkNisklLTVOHl1bPwWIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false
```
interface Foo {
  bar: number;
}

function main(foo: Foo | null, foos: Foo[]): boolean {
  if (foo && foos[0] && foo.bar === foos[0].bar) {
    return true
  }
  return false
}
```

**Comment by Hypnosphi** [comment]
Ok, I think I know where the confusion comes from. In js, `===` has higher precedence than `&&`, so `foo && foo.bar === bar?.bar` is equivalent to `foo && (foo.bar === bar?.bar)` rather than `(foo && foo.bar) === bar?.bar`. 

Here's the same example with parens for clarity, giving the same invalid fix:
https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyARnFAFzIgCuAtsdANwEC%2BBBM1ICYw6IyWnFAAKGJkoYsAHyrUANvIA0ycegDOkzAG0AugEpKxTPIhwBeQsmAxkYzMgBkju2oB0pKMgC8v1ZnVtAAZdAH4PMn19HAIiIigIMGooATAoaghY5DYrBKSU1Th5dUz8FiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg10NM8AOXapUAYQAW6aAGsylDul4BfECqA&tokens=false

**Comment by JEmilyKoo** [comment]
I'd like to work on this. I will open a draft PR soon.

**Comment by kirkwaiblinger** [comment]
This is more or less a variant of https://github.com/typescript-eslint/typescript-eslint/issues/11840. Note that the rule does currently seem to try to be smart about whether the right hand side of the comparison is `undefined`, but it's obscured from the type system by the array indexed access producing non-undefined type. Enabling `noUncheckedIndexedAcccess: true`, or using `.at()` here both prevent the rule from autofixing.

I personally don't like the `a && a.b === c` => `a?.b === c` report at all because it has confusing edge cases and limitations such as this one. I agree with at least downgrading to a suggestion, but I really would prefer that the whole pattern be guarded behind an option.

**Comment by Hypnosphi** [comment]
Such an autofix would be totally fine if the right hand was a literal, see "A slightly longer example" from #7170

```
foo && foo.bar && foo.bar.baz === 1
```

**Comment by kirkwaiblinger** [comment]
Coming back to this, I think that the fix of turning all `a && a.b === c` => `a?.b === c` fixes into suggestions is the right pragmatic approach for the time being. 

@Hypnosphi I think that https://github.com/typescript-eslint/typescript-eslint/issues/11917#issuecomment-3834546819, about literals, is valid but makes sense to be considered as a feature request after resolution of this issue, not a blocker for actioning this issue. 

Fundamentally, we're being (much) too aggressive right now with the `a && a.b === c` => `a?.b === c` case. Let's take that down a notch with a broad brush, before considering if there are special cases that are worth bringing back where the aggression _is_ safe.

**Comment by bug-origin** [comment]
I'll work on fixing this issue. The problem is that the auto-fix changes the semantics when comparing with optional chaining on the right side. I'll investigate the fixer logic and submit a PR.
